src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.CompletionListUpdater.cs

```diff
@@ -0,0 +1,805 @@
+            private static readonly ObjectPool<List<MatchResult<VSCompletionItem>>> s_listOfMatchResultPool = new(factory: () => new(), size: 1);
+
+            public CompletionListUpdater(
+                IAsyncCompletionSession session,
+                AsyncCompletionSessionDataSnapshot data,
+                RecentItemsManager recentItemsManager,
+                IGlobalOptionService globalOptions)
+            {
+                _session = session;
+                _data = data;
+                _recentItemsManager = recentItemsManager;
+
+                _filterText = _session.ApplicableToSpan.GetText(_data.Snapshot);
+
+                if (!_session.Properties.TryGetProperty(CompletionSource.HasSuggestionItemOptions, out bool hasSuggestedItemOptions))
+                {
+                    // This is the scenario when the session is created out of Roslyn, in some other provider, e.g. in Debugger.
+                    // For now, the default hasSuggestedItemOptions is false.
+                    hasSuggestedItemOptions = false;
+                }
+
+                _hasSuggestedItemOptions = hasSuggestedItemOptions || _data.DisplaySuggestionItem;
+
+                // We prefer using the original snapshot, which should always be available from items provided by Roslyn's CompletionSource.
+                // Only use data.Snapshot in the theoretically possible but rare case when all items we are handling are from some non-Roslyn CompletionSource.
+                var snapshotForDocument = TryGetInitialTriggerLocation(_data, out var intialTriggerLocation)
+                    ? intialTriggerLocation.Snapshot
+                    : _data.Snapshot;
+
+                _document = snapshotForDocument?.TextBuffer.AsTextContainer().GetOpenDocumentInCurrentContext();
+                _completionService = _document?.GetLanguageService<CompletionService>();
+                _completionRules = _completionService?.GetRules(CompletionOptions.From(_document!.Project)) ?? CompletionRules.Default;
+
+                // Let us make the completion Helper used for non-Roslyn items case-sensitive.
+                // We can change this if get requests from partner teams.
+                _completionHelper = _document != null ? CompletionHelper.GetHelper(_document) : new CompletionHelper(isCaseSensitive: true);
+                _filterMethod = _completionService == null
+                    ? ((itemsWithPatternMatches, text) => CompletionService.FilterItems(_completionHelper, itemsWithPatternMatches, text))
+                    : ((itemsWithPatternMatches, text) => _completionService.FilterItems(_document!, itemsWithPatternMatches, text));
+
+                // Nothing to highlight if user hasn't typed anything yet.
+                _highlightMatchingPortions = _filterText.Length > 0
+                        && globalOptions.GetOption(CompletionViewOptions.HighlightMatchingPortionsOfCompletionListItems, _document?.Project.Language);
```

Comment:
```suggestion
                _highlightMatchingPortions = _filterText.Length > 0
                    && globalOptions.GetOption(CompletionViewOptions.HighlightMatchingPortionsOfCompletionListItems, _document?.Project.Language);
```

in general, when doing indentation, try to either:

1. indent 4 spaces from teh parent
2. indent such that the child is aligned with something.